### PR TITLE
[Report Page] Add caching to new report page

### PR DESCRIPTION
**Pull request by j-x-han**
# Description

Adds caching for the new pipeline report page.

Follows the caching pattern used in the previous report: once a pipeline run has finished, reports against all backgrounds will be precached.

# Notes

`PipelineReportService.call` was modified to take a `pipeline_run` instead of a `pipeline_run.id` to reduce some redundancy in the code. Calls to the function were updated accordingly.

When porting the cache-related code in `report_info` into `report_v2`, calls to `MetricUtil.put_metric_now` were replaced with `MetricUtil.log_analytics_event`.

# Tests

* Enabled caching in development and verified that the report information would be fetched from the cache rather than re-computed from scratch.

### CODE REVIEW

File: app/controllers/samples_controller.rb [hunk 1]
```diff
@@ -800,19 +800,39 @@ def samples_going_public
...
-          jobStatus: "Waiting to Start or Receive Files",
-          errorMessage: nil,
-          knownUserError: nil,
-        },
-      }
-    end
+    background_id = get_background_id(@sample, params[:background])
+
+    skip_cache = params[:skip_cache] || false
+    MetricUtil.log_analytics_event("samples.cache.requested", current_user) unless skip_cache
```
Comment (jshoe) [inline app/controllers/samples_controller.rb]: For the `log_analytics_event` calls, could you changes the names to the other format? https://czi.quip.com/67RCAIiHN0Qc/IDseq-product-analytics-How-to-log

File: app/models/pipeline_run.rb [hunk 2]
```diff
@@ -1577,6 +1578,22 @@ def precache_report_info!
...
+      cache_key = PipelineReportService.report_info_cache_key(
+        "/samples/#{sample.id}/report_v2",
+        params.merge(background_id: background_id)
+      )
+      Rails.logger.info("Precaching #{cache_key} with background #{background_id}")
+      Rails.cache.fetch(cache_key, expires_in: 30.days) do
+        PipelineReportService.call(self, background_id)
+      end
+
+      MetricUtil.log_analytics_event("samples.cache.precached", current_user)
```
Comment (jshoe) [inline app/models/pipeline_run.rb]: also not sure who `current_user` would be in this example, can probably leave it blank

File: app/controllers/samples_controller.rb [hunk 3]
```diff
@@ -800,19 +800,39 @@ def samples_going_public
...
+    report_info_params = pipeline_run.report_info_params
+
+    cache_key = PipelineReportService.report_info_cache_key(
+      request.path,
+      params
+        .permit(report_info_params.keys)
+        .merge(pipeline_run_id: pipeline_run.id)
+    )
+
+    json =
```
Comment (jshoe) [inline app/controllers/samples_controller.rb]: It feels hard to read this conditional assignment with so much in the `else` block. did Rubocop force this? Maybe have it as something like `unless skip_cache, compute all the headers`?

File: app/controllers/samples_controller.rb (hunk 1 above)
Comment (j-x-han) [reply app/controllers/samples_controller.rb]: Changed!

File: app/models/pipeline_run.rb (hunk 2 above)
Comment (j-x-han) [reply app/models/pipeline_run.rb]: `log_analytics_event` requires a user to be passed in, so I replaced it with `nil`.

File: app/controllers/samples_controller.rb (hunk 3 above)
Comment (j-x-han) [reply app/controllers/samples_controller.rb]: At @tfrcarvalho 's request, I pulled the cache logic out into its own function so the block should be cleaner now!

File: app/services/pipeline_report_service.rb [hunk 4]
```diff
@@ -67,12 +67,13 @@ def call
   end
 
   def generate
-    pipeline_run, metadata = get_pipeline_status(@pipeline_run_id)
+    pipeline_run, metadata = get_pipeline_status(@pipeline_run)
```
Comment (tfrcarvalho) [inline app/services/pipeline_report_service.rb]: This makes no sense if you are sending in a pipeline_run now. We are returning a pipeline run and sending it as input?

File: app/controllers/samples_controller.rb [hunk 5]
```diff
@@ -800,19 +800,23 @@ def samples_going_public
...
+    cache_key = PipelineReportService.report_info_cache_key(
+      request.path,
+      params
+        .permit(report_info_params.keys)
+        .merge(pipeline_run_id: pipeline_run.id)
+    )
+    httpdate = Time.at(report_info_params[:report_ts]).utc.httpdate
+
+    json =
+      fetch_from_cache(skip_cache, cache_key, httpdate, "PipelineReport") do
```
Comment (tfrcarvalho) [inline app/controllers/samples_controller.rb]: `fetch_from_or_store_in_cache`?

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -173,4 +173,25 @@ def instrument_with_timer
     yield
     @timer.publish
   end
+
+  def fetch_from_cache(skip_cache, cache_key, httpdate, event_name)
```
Comment (tfrcarvalho) [inline app/controllers/application_controller.rb]: Add a comment explaining this method.

File: app/services/pipeline_report_service.rb (hunk 4 above)
Comment (j-x-han) [reply app/services/pipeline_report_service.rb]: Removed pipeline_run as an output.

File: app/controllers/samples_controller.rb (hunk 5 above)
Comment (j-x-han) [reply app/controllers/samples_controller.rb]: Done.

File: app/controllers/application_controller.rb (hunk 6 above)
Comment (j-x-han) [reply app/controllers/application_controller.rb]: Done.

**Pull request merged**